torch/optim/lr_scheduler.py

```diff
@@ -427,3 +427,83 @@ def state_dict(self):
+    r"""Set the learning rate of each parameter group using a cosine annealing
+    schedule, where :math:`\eta_{max}` is set to the initial lr, :math:`T_{cur}`
+    is the number of epochs since the last restart and :math:`T_{i}` is the number
+    of epochs between two warm restarts in SGDR:
+
+    .. math::
+
+        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
+        \cos(\frac{T_{cur}}{T_{i}}\pi))
+
+    When :math:`\T_{cur}=T_{i}`, set :math:`\eta_t = \eta_{min}`.
+    When :math:`\T_{cur}=0`(after restart), set :math:`\eta_t=\eta_{max}`.
+
+    It has been proposed in
+    `SGDR: Stochastic Gradient Descent with Warm Restarts`_.
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        T_0 (int): Number of iterations for the first restart.
+        T_mult (float): A factor increases :math:`\T_{i}` after a restart. Default: 1.
+        eta_min (float): Minimum learning rate. Default: 0.
+        last_epoch (int): The index of last epoch. Default: -1.
+
+    .. _SGDR\: Stochastic Gradient Descent with Warm Restarts:
+        https://arxiv.org/abs/1608.03983
+    """
+
+    def __init__(self, optimizer, T_0, T_mult=1, eta_min=0, last_epoch=-1):
+        T_0 = int(T_0)
+        if T_0 <= 0:
+            raise ValueError("Expected positive T_0, but got {}".format(T_0))
+        if T_mult < 1:
+            raise ValueError("Expected T_mul >= 1, but got {}".format(T_mul))
+        self.T_0 = T_0
+        self.T_i = T_0
+        self.T_mult = T_mult
+        self.eta_min = eta_min
+        super(SGDR, self).__init__(optimizer, last_epoch)
+        self.T_cur = last_epoch
+
+    def get_lr(self):
+        return [self.eta_min + (base_lr - self.eta_min) * (1 + math.cos(math.pi * self.T_cur / self.T_i)) / 2
+                for base_lr in self.base_lrs]
+
+    def step(self, epoch=None):
+        """Step could be called after every update, i.e. if one epoch has 10 iterations
+        (number_of_train_examples / batch_size), we should call SGDR.step(0.1), SGDR.step(0.2), etc.
+
+        This function can be called in an interleaved way.
+
+        Example:
+            >>> scheduler = SGDR(optimizer, T_0, T_mult)
+            >>> for epoch in range(20):
+            >>>     scheduler.step()
+            >>> scheduler.step(26)
+            >>> scheduler.step() # scheduler.step(27), instead of scheduler(20)
+        """
+        if epoch is None:
+            epoch = self.last_epoch + 1
+            self.T_cur = self.T_cur + 1
+            if self.T_cur >= self.T_i:
+                self.T_cur = self.T_cur - self.T_i
+                self.T_i = self.T_i * self.T_mult
+        else:
+            if epoch >= self.T_0:
+                if self.T_mult == 1:
+                    self.T_cur = epoch % self.T_0
+                else:
+                    n = int(math.log((epoch / self.T_0 * (self.T_mult - 1) + 1), self.T_mult))
+                    self.T_cur = epoch - self.T_0 * (self.T_mult ** n - 1) / (self.T_mult - 1)
```

Comment:
It could be my suggestion introduced numerical instability. It makes me rethink whether it makes sense to have non-integer `T_i`. Any thought? @Kirayue

Reply:
It is not restricted to the integer from my opinion and the origin paper. But if we can make sure that the failure is caused by numerical instability, I can modify the code. 

By the way, how can I reproduce the failure? (I tested it in python3, it works.)

Reply:
Hi @yf225 is there a way to reproduce CI test failures for a specific environment? It does not have to be local, as long as @Kirayue can iterate with that conveniently.

Reply:
Hi @Kirayue, please feel free to use the instructions at https://gist.github.com/yf225/9af6f95efd0923b08ba7733611d80d57 to reproduce CI test failures in your local environment. (We clean up unused Docker images periodically, and you might need to rerun the CI and then pull the Docker image as soon as it fails to keep them in your local environment.)

Reply:
@mrshenli Sorry for the unfortunate update: I just found out that currently we don't have a way to make the CI Docker images public, and we cannot share the credentials to access the Docker images in ECR either. I am working on a solution at https://github.com/pytorch/pytorch/pull/18244.

Reply:
@yf225 Thanks for letting us know.

Reply:
@Kirayue sorry for the delay on this - would you like to send us your email address so that we can share the credentials for read-only access to the Docker images in ECR? Thanks!

Reply:
Do not worry about it :smile: 
Sure, just use this email address `kirayue1021@gmail.com`